// buefy/dist/cjs/helpers.js
'use strict';

var helpers = require('./helpers-bVbODwYf.js');
require('./typeof-fhICEa4F.js');
require('vue');



exports.bound = helpers.bound;
exports.copyAppContext = helpers.copyAppContext;
exports.createAbsoluteElement = helpers.createAbsoluteElement;
exports.createNewEvent = helpers.createNewEvent;
exports.escapeRegExpChars = helpers.escapeRegExpChars;
exports.getComponentFromVNode = helpers.getComponentFromVNode;
exports.getMonthNames = helpers.getMonthNames;
exports.getValueByPath = helpers.getValueByPath;
exports.getWeekdayNames = helpers.getWeekdayNames;
exports.hasFlag = helpers.hasFlag;
exports.indexOf = helpers.indexOf;
exports.isCustomElement = helpers.isCustomElement;
exports.isDefined = helpers.isDefined;
exports.isFragment = helpers.isFragment;
exports.isMobile = helpers.isMobile;
exports.isNil = helpers.isNil;
exports.isTag = helpers.isTag;
exports.isVueComponent = helpers.isVueComponent;
exports.isWebpSupported = helpers.isWebpSupported;
exports.matchWithGroups = helpers.matchWithGroups;
exports.merge = helpers.merge;
exports.mod = helpers.mod;
exports.multiColumnSort = helpers.multiColumnSort;
exports.removeDiacriticsFromString = helpers.removeDiacriticsFromString;
exports.removeElement = helpers.removeElement;
exports.sign = helpers.sign;
exports.toCssWidth = helpers.toCssWidth;
exports.translateTouchAsDragEvent = helpers.translateTouchAsDragEvent;
