// buefy/dist/cjs/Select-toBXychX.js
'use strict';

var Icon = require('./Icon-o1NBLDNV.js');
var CompatFallthroughMixin = require('./CompatFallthroughMixin-ZOMao9mo.js');
var FormElementMixin = require('./FormElementMixin-RZ3syZQl.js');
var vue = require('vue');

var script = {
    name: 'BSelect',
    components: {
        [Icon.script.name]: Icon.script
    },
    mixins: [CompatFallthroughMixin.CompatFallthroughMixin, FormElementMixin.FormElementMixin],
    props: {
        modelValue: {
            type: [String, Number, Boolean, Object, Array, Function, Date],
            default: null
        },
        placeholder: String,
        multiple: Boolean,
        nativeSize: [String, Number]
    },
    emits: ['blur', 'focus', 'update:modelValue'],
    data() {
        return {
            selected: this.modelValue,
            _elementRef: 'select'
        }
    },
    computed: {
        computedValue: {
            get() {
                return this.selected
            },
            set(value) {
                this.selected = value;
                this.$emit('update:modelValue', value);
                !this.isValid && this.checkHtml5Validity();
            }
        },
        spanClasses() {
            return [this.size, this.statusType, {
                'is-fullwidth': this.expanded,
                'is-loading': this.loading,
                'is-multiple': this.multiple,
                'is-rounded': this.rounded,
                'is-empty': this.selected === null
            }]
        }
    },
    watch: {
        /**
        * When v-model is changed:
        *   1. Set the selected option.
        *   2. If it's invalid, validate again.
        */
        modelValue(value) {
            this.selected = value;
            !this.isValid && this.checkHtml5Validity();
        }
    }
};

const _hoisted_1 = ["multiple", "size"];
const _hoisted_2 = {
  key: 0,
  value: null,
  disabled: "",
  hidden: ""
};

function render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_b_icon = vue.resolveComponent("b-icon");

  return (vue.openBlock(), vue.createElementBlock("div", vue.mergeProps({
    class: ["control", { 'is-expanded': _ctx.expanded, 'has-icons-left': _ctx.icon }]
  }, _ctx.rootAttrs), [
    vue.createElementVNode("span", {
      class: vue.normalizeClass(["select", $options.spanClasses])
    }, [
      vue.withDirectives(vue.createElementVNode("select", vue.mergeProps({
        "onUpdate:modelValue": _cache[0] || (_cache[0] = $event => (($options.computedValue) = $event)),
        ref: "select",
        multiple: $props.multiple,
        size: $props.nativeSize
      }, _ctx.fallthroughAttrs, {
        onBlur: _cache[1] || (_cache[1] = $event => (_ctx.$emit('blur', $event) && _ctx.checkHtml5Validity())),
        onFocus: _cache[2] || (_cache[2] = $event => (_ctx.$emit('focus', $event)))
      }), [
        ($props.placeholder)
          ? (vue.openBlock(), vue.createElementBlock(vue.Fragment, { key: 0 }, [
              ($options.computedValue == null)
                ? (vue.openBlock(), vue.createElementBlock("option", _hoisted_2, vue.toDisplayString($props.placeholder), 1 /* TEXT */))
                : vue.createCommentVNode("v-if", true)
            ], 64 /* STABLE_FRAGMENT */))
          : vue.createCommentVNode("v-if", true),
        vue.renderSlot(_ctx.$slots, "default")
      ], 16 /* FULL_PROPS */, _hoisted_1), [
        [vue.vModelSelect, $options.computedValue]
      ])
    ], 2 /* CLASS */),
    (_ctx.icon)
      ? (vue.openBlock(), vue.createBlock(_component_b_icon, {
          key: 0,
          class: "is-left",
          icon: _ctx.icon,
          pack: _ctx.iconPack,
          size: _ctx.iconSize
        }, null, 8 /* PROPS */, ["icon", "pack", "size"]))
      : vue.createCommentVNode("v-if", true)
  ], 16 /* FULL_PROPS */))
}

script.render = render;
script.__file = "src/components/select/Select.vue";

exports.script = script;
