// buefy/dist/components/progress/index.js
/*! Buefy v0.1.3 | MIT License | github.com/buefy/buefy */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('vue')) :
    typeof define === 'function' && define.amd ? define(['exports', 'vue'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.Progress = {}, global.Vue));
})(this, (function (exports, vue) { 'use strict';

    var config = {
      defaultContainerElement: null,
      defaultIconPack: 'mdi',
      defaultIconComponent: null,
      defaultIconPrev: 'chevron-left',
      defaultIconNext: 'chevron-right',
      defaultLocale: undefined,
      defaultDialogConfirmText: null,
      defaultDialogCancelText: null,
      defaultSnackbarDuration: 3500,
      defaultSnackbarPosition: null,
      defaultToastDuration: 2000,
      defaultToastPosition: null,
      defaultNotificationDuration: 2000,
      defaultNotificationPosition: null,
      defaultTooltipType: 'is-primary',
      defaultTooltipDelay: null,
      defaultTooltipCloseDelay: null,
      defaultSidebarDelay: null,
      defaultInputAutocomplete: 'on',
      defaultDateFormatter: null,
      defaultDateParser: null,
      defaultDateCreator: null,
      defaultTimeCreator: null,
      defaultDayNames: null,
      defaultMonthNames: null,
      defaultFirstDayOfWeek: null,
      defaultUnselectableDaysOfWeek: null,
      defaultTimeFormatter: null,
      defaultTimeParser: null,
      defaultModalCanCancel: ['escape', 'x', 'outside', 'button'],
      defaultModalScroll: null,
      defaultDatepickerMobileNative: true,
      defaultTimepickerMobileNative: true,
      defaultTimepickerMobileModal: true,
      defaultNoticeQueue: true,
      defaultInputHasCounter: true,
      /**
       * Whether `class`, `style`, and `id` are applied to the root element in
       * components that are affected by Vue 3 change in fallthgourh beahvior.
       * See: https://github.com/ntohq/buefy-next/issues/16
       */
      defaultCompatFallthrough: true,
      defaultTaginputHasCounter: true,
      defaultUseHtml5Validation: true,
      defaultDropdownMobileModal: true,
      defaultFieldLabelPosition: null,
      defaultDatepickerYearsRange: [-100, 10],
      defaultDatepickerNearbyMonthDays: true,
      defaultDatepickerNearbySelectableMonthDays: false,
      defaultDatepickerShowWeekNumber: false,
      defaultDatepickerWeekNumberClickable: false,
      defaultDatepickerMobileModal: true,
      defaultTrapFocus: true,
      defaultAutoFocus: true,
      defaultButtonRounded: false,
      defaultSwitchRounded: true,
      defaultCarouselInterval: 3500,
      defaultTabsExpanded: false,
      defaultTabsAnimated: true,
      defaultTabsType: null,
      defaultStatusIcon: true,
      defaultProgrammaticPromise: false,
      defaultLinkTags: ['a', 'button', 'input', 'router-link', 'nuxt-link', 'n-link', 'RouterLink', 'NuxtLink', 'NLink'],
      defaultImageWebpFallback: null,
      defaultImageLazy: true,
      defaultImageResponsive: true,
      defaultImageRatio: null,
      defaultImageSrcsetFormatter: null,
      defaultBreadcrumbTag: 'a',
      defaultBreadcrumbAlign: 'is-left',
      defaultBreadcrumbSeparator: '',
      defaultBreadcrumbSize: 'is-medium',
      customIconPacks: null
    };

    const PROGRESS_INJECTION_KEY = Symbol('bprogress');

    var script$1 = {
        name: 'BProgress',
        provide() {
            return {
                [PROGRESS_INJECTION_KEY]: this
            }
        },
        props: {
            type: {
                type: [String, Object],
                default: 'is-darkgrey'
            },
            size: String,
            rounded: {
                type: Boolean,
                default: true
            },
            value: {
                type: Number,
                default: undefined
            },
            max: {
                type: Number,
                default: 100
            },
            showValue: {
                type: Boolean,
                default: false
            },
            format: {
                type: String,
                default: 'raw',
                validator: (value) => {
                    return [
                        'raw',
                        'percent'
                    ].indexOf(value) >= 0
                }
            },
            precision: {
                type: Number,
                default: 2
            },
            keepTrailingZeroes: {
                type: Boolean,
                default: false
            },
            locale: {
                type: [String, Array],
                default: () => {
                    return config.defaultLocale
                }
            }
        },
        computed: {
            isIndeterminate() {
                return this.value === undefined || this.value === null
            },
            newType() {
                return [
                    this.size,
                    this.type,
                    {
                        'is-more-than-half': this.value && this.value > this.max / 2
                    }
                ]
            },
            newValue() {
                return this.calculateValue(this.value)
            },
            isNative() {
                return this.$slots.bar === undefined
            },
            wrapperClasses() {
                return {
                    'is-not-native': !this.isNative,
                    [this.size]: typeof this.size === 'string' && !this.isNative
                }
            }
        },
        watch: {
            /**
             * When value is changed back to undefined, value of native progress get reset to 0.
             * Need to add and remove the value attribute to have the indeterminate or not.
             */
            isIndeterminate(indeterminate) {
                this.$nextTick(() => {
                    if (this.$refs.progress) {
                        if (indeterminate) {
                            this.$refs.progress.removeAttribute('value');
                        } else {
                            this.$refs.progress.setAttribute('value', this.value);
                        }
                    }
                });
            }
        },
        methods: {
            calculateValue(value) {
                if (value === undefined || value === null || isNaN(value)) {
                    return undefined
                }

                const minimumFractionDigits = this.keepTrailingZeroes ? this.precision : 0;
                const maximumFractionDigits = this.precision;
                if (this.format === 'percent') {
                    return new Intl.NumberFormat(
                        this.locale,
                        {
                            style: 'percent',
                            minimumFractionDigits,
                            maximumFractionDigits
                        }
                    ).format(value / this.max)
                }

                return new Intl.NumberFormat(
                    this.locale,
                    {
                        minimumFractionDigits,
                        maximumFractionDigits
                    }
                ).format(value)
            }
        }
    };

    const _hoisted_1$1 = ["max", "value"];
    const _hoisted_2$1 = {
      key: 2,
      class: "progress-value"
    };

    function render$1(_ctx, _cache, $props, $setup, $data, $options) {
      return (vue.openBlock(), vue.createElementBlock("div", {
        class: vue.normalizeClass(["progress-wrapper", [$options.wrapperClasses, { 'is-squared': !$props.rounded }]])
      }, [
        ($options.isNative)
          ? (vue.openBlock(), vue.createElementBlock("progress", {
              key: 0,
              ref: "progress",
              class: vue.normalizeClass(["progress", [$options.newType, { 'is-squared': !$props.rounded }]]),
              max: $props.max,
              value: $props.value
            }, vue.toDisplayString($options.newValue), 11 /* TEXT, CLASS, PROPS */, _hoisted_1$1))
          : vue.renderSlot(_ctx.$slots, "bar", { key: 1 }),
        ($options.isNative && $props.showValue)
          ? (vue.openBlock(), vue.createElementBlock("p", _hoisted_2$1, [
              vue.renderSlot(_ctx.$slots, "default", {}, () => [
                vue.createTextVNode(vue.toDisplayString($options.newValue), 1 /* TEXT */)
              ])
            ]))
          : vue.createCommentVNode("v-if", true)
      ], 2 /* CLASS */))
    }

    script$1.render = render$1;
    script$1.__file = "src/components/progress/Progress.vue";

    var script = {
        name: 'BProgressBar',
        inject: {
            parent: {
                from: PROGRESS_INJECTION_KEY,
                default: undefined
            }
        },
        props: {
            type: {
                type: [String, Object],
                default: undefined
            },
            value: {
                type: Number,
                default: undefined
            },
            showValue: {
                type: Boolean,
                default: false
            }
        },
        computed: {
            newType() {
                return [
                    this.parent.size,
                    this.type || this.parent.type
                ]
            },
            newShowValue() {
                return this.showValue || this.parent.showValue
            },
            newValue() {
                return this.parent.calculateValue(this.value)
            },
            barWidth() {
                return `${this.value * 100 / this.parent.max}%`
            }
        }
    };

    const _hoisted_1 = ["aria-valuenow", "aria-valuemax"];
    const _hoisted_2 = {
      key: 0,
      class: "progress-value"
    };

    function render(_ctx, _cache, $props, $setup, $data, $options) {
      return (vue.openBlock(), vue.createElementBlock("div", {
        class: vue.normalizeClass(["progress-bar", $options.newType]),
        role: "progressbar",
        "aria-valuenow": $props.value,
        "aria-valuemax": $options.parent.max,
        "aria-valuemin": "0",
        style: vue.normalizeStyle({width: $options.barWidth})
      }, [
        ($options.newShowValue)
          ? (vue.openBlock(), vue.createElementBlock("p", _hoisted_2, [
              vue.renderSlot(_ctx.$slots, "default", {}, () => [
                vue.createTextVNode(vue.toDisplayString($options.newValue), 1 /* TEXT */)
              ])
            ]))
          : vue.createCommentVNode("v-if", true)
      ], 14 /* CLASS, STYLE, PROPS */, _hoisted_1))
    }

    script.render = render;
    script.__file = "src/components/progress/ProgressBar.vue";

    var use = function use(plugin) {
      if (typeof window !== 'undefined' && window.Vue) {
        window.Vue.use(plugin);
      }
    };

    // use `name` to register a Functional Component which will become unresolvable
    // in production build due to name mangling.
    var registerComponent = function registerComponent(Vue, component, name) {
      Vue.component(name || component.name, component);
    };

    var Plugin = {
      install: function install(Vue) {
        registerComponent(Vue, script$1);
        registerComponent(Vue, script);
      }
    };
    use(Plugin);

    exports.BProgress = script$1;
    exports.BProgressBar = script;
    exports.default = Plugin;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
