// buefy/dist/components/menu/index.js
/*! Buefy v0.1.3 | MIT License | github.com/buefy/buefy */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('vue')) :
    typeof define === 'function' && define.amd ? define(['exports', 'vue'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.Menu = {}, global.Vue));
})(this, (function (exports, vue) { 'use strict';

    var MenuItemContainerMixin = {
      provide: function provide() {
        return {
          BMenuItemContainer: this
        };
      },
      data: function data() {
        return {
          menuItems: []
        };
      },
      methods: {
        appendMenuItem: function appendMenuItem(item) {
          this.menuItems.push(item);
        },
        removeMenuItem: function removeMenuItem(item) {
          var index = this.menuItems.indexOf(item);
          if (index !== -1) {
            this.menuItems.splice(index, 1);
          }
        }
      }
    };

    var script$3 = {
        name: 'BMenu',
        mixins: [MenuItemContainerMixin],
        props: {
            accordion: {
                type: Boolean,
                default: true
            },
            activable: {
                type: Boolean,
                default: true
            }
        },
        data() {
            return {
                _isMenu: true // Used by MenuItem
            }
        }
    };

    const _hoisted_1$2 = { class: "menu" };

    function render$3(_ctx, _cache, $props, $setup, $data, $options) {
      return (vue.openBlock(), vue.createElementBlock("div", _hoisted_1$2, [
        vue.renderSlot(_ctx.$slots, "default")
      ]))
    }

    script$3.render = render$3;
    script$3.__file = "src/components/menu/Menu.vue";

    var script$2 = {
        name: 'BMenuList',
        props: {
            label: String,
            icon: String,
            iconPack: String,
            ariaRole: {
                type: String,
                default: ''
            },
            size: {
                type: String,
                default: 'is-small'
            }
        }
    };

    const _hoisted_1$1 = {
      key: 0,
      class: "menu-label"
    };
    const _hoisted_2$1 = ["role"];

    function render$2(_ctx, _cache, $props, $setup, $data, $options) {
      const _component_b_icon = vue.resolveComponent("b-icon");

      return (vue.openBlock(), vue.createElementBlock(vue.Fragment, null, [
        ($props.label || _ctx.$slots.label)
          ? (vue.openBlock(), vue.createElementBlock("p", _hoisted_1$1, [
              ($props.label)
                ? (vue.openBlock(), vue.createElementBlock(vue.Fragment, { key: 0 }, [
                    ($props.icon)
                      ? (vue.openBlock(), vue.createElementBlock(vue.Fragment, { key: 0 }, [
                          vue.createVNode(_component_b_icon, {
                            icon: $props.icon,
                            pack: $props.iconPack,
                            size: $props.size
                          }, null, 8 /* PROPS */, ["icon", "pack", "size"]),
                          vue.createElementVNode("span", null, vue.toDisplayString($props.label), 1 /* TEXT */)
                        ], 64 /* STABLE_FRAGMENT */))
                      : (vue.openBlock(), vue.createElementBlock(vue.Fragment, { key: 1 }, [
                          vue.createTextVNode(vue.toDisplayString($props.label), 1 /* TEXT */)
                        ], 64 /* STABLE_FRAGMENT */))
                  ], 64 /* STABLE_FRAGMENT */))
                : vue.renderSlot(_ctx.$slots, "label", { key: 1 })
            ]))
          : vue.createCommentVNode("v-if", true),
        vue.createElementVNode("ul", {
          class: "menu-list",
          role: $props.ariaRole === 'menu' ? $props.ariaRole : undefined
        }, [
          vue.renderSlot(_ctx.$slots, "default")
        ], 8 /* PROPS */, _hoisted_2$1)
      ], 64 /* STABLE_FRAGMENT */))
    }

    script$2.render = render$2;
    script$2.__file = "src/components/menu/MenuList.vue";

    var config = {
      defaultContainerElement: null,
      defaultIconPack: 'mdi',
      defaultIconComponent: null,
      defaultIconPrev: 'chevron-left',
      defaultIconNext: 'chevron-right',
      defaultLocale: undefined,
      defaultDialogConfirmText: null,
      defaultDialogCancelText: null,
      defaultSnackbarDuration: 3500,
      defaultSnackbarPosition: null,
      defaultToastDuration: 2000,
      defaultToastPosition: null,
      defaultNotificationDuration: 2000,
      defaultNotificationPosition: null,
      defaultTooltipType: 'is-primary',
      defaultTooltipDelay: null,
      defaultTooltipCloseDelay: null,
      defaultSidebarDelay: null,
      defaultInputAutocomplete: 'on',
      defaultDateFormatter: null,
      defaultDateParser: null,
      defaultDateCreator: null,
      defaultTimeCreator: null,
      defaultDayNames: null,
      defaultMonthNames: null,
      defaultFirstDayOfWeek: null,
      defaultUnselectableDaysOfWeek: null,
      defaultTimeFormatter: null,
      defaultTimeParser: null,
      defaultModalCanCancel: ['escape', 'x', 'outside', 'button'],
      defaultModalScroll: null,
      defaultDatepickerMobileNative: true,
      defaultTimepickerMobileNative: true,
      defaultTimepickerMobileModal: true,
      defaultNoticeQueue: true,
      defaultInputHasCounter: true,
      /**
       * Whether `class`, `style`, and `id` are applied to the root element in
       * components that are affected by Vue 3 change in fallthgourh beahvior.
       * See: https://github.com/ntohq/buefy-next/issues/16
       */
      defaultCompatFallthrough: true,
      defaultTaginputHasCounter: true,
      defaultUseHtml5Validation: true,
      defaultDropdownMobileModal: true,
      defaultFieldLabelPosition: null,
      defaultDatepickerYearsRange: [-100, 10],
      defaultDatepickerNearbyMonthDays: true,
      defaultDatepickerNearbySelectableMonthDays: false,
      defaultDatepickerShowWeekNumber: false,
      defaultDatepickerWeekNumberClickable: false,
      defaultDatepickerMobileModal: true,
      defaultTrapFocus: true,
      defaultAutoFocus: true,
      defaultButtonRounded: false,
      defaultSwitchRounded: true,
      defaultCarouselInterval: 3500,
      defaultTabsExpanded: false,
      defaultTabsAnimated: true,
      defaultTabsType: null,
      defaultStatusIcon: true,
      defaultProgrammaticPromise: false,
      defaultLinkTags: ['a', 'button', 'input', 'router-link', 'nuxt-link', 'n-link', 'RouterLink', 'NuxtLink', 'NLink'],
      defaultImageWebpFallback: null,
      defaultImageLazy: true,
      defaultImageResponsive: true,
      defaultImageRatio: null,
      defaultImageSrcsetFormatter: null,
      defaultBreadcrumbTag: 'a',
      defaultBreadcrumbAlign: 'is-left',
      defaultBreadcrumbSeparator: '',
      defaultBreadcrumbSize: 'is-medium',
      customIconPacks: null
    };

    function _typeof(o) {
      "@babel/helpers - typeof";

      return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) {
        return typeof o;
      } : function (o) {
        return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
      }, _typeof(o);
    }

    function _toPrimitive(input, hint) {
      if (_typeof(input) !== "object" || input === null) return input;
      var prim = input[Symbol.toPrimitive];
      if (prim !== undefined) {
        var res = prim.call(input, hint || "default");
        if (_typeof(res) !== "object") return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
      }
      return (hint === "string" ? String : Number)(input);
    }

    function _toPropertyKey(arg) {
      var key = _toPrimitive(arg, "string");
      return _typeof(key) === "symbol" ? key : String(key);
    }

    function _defineProperty(obj, key, value) {
      key = _toPropertyKey(key);
      if (key in obj) {
        Object.defineProperty(obj, key, {
          value: value,
          enumerable: true,
          configurable: true,
          writable: true
        });
      } else {
        obj[key] = value;
      }
      return obj;
    }

    function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
    function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }

    /**
     * Merge function to replace Object.assign with deep merging possibility
     */
    var isObject = function isObject(item) {
      return _typeof(item) === 'object' && !Array.isArray(item);
    };
    var mergeFn = function mergeFn(target, source) {
      var deep = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
      if (deep || !Object.assign) {
        var isDeep = function isDeep(prop) {
          return isObject(source[prop]) && target !== null && Object.prototype.hasOwnProperty.call(target, prop) && isObject(target[prop]);
        };
        var replaced = Object.getOwnPropertyNames(source).map(function (prop) {
          return _defineProperty({}, prop, isDeep(prop) ? mergeFn(target[prop], source[prop], deep) : source[prop]);
        }).reduce(function (a, b) {
          return _objectSpread(_objectSpread({}, a), b);
        }, {});
        return _objectSpread(_objectSpread({}, target), replaced);
      } else {
        return Object.assign(target, source);
      }
    };
    var merge = mergeFn;

    var mdiIcons = {
      sizes: {
        "default": 'mdi-24px',
        'is-small': null,
        'is-medium': 'mdi-36px',
        'is-large': 'mdi-48px'
      },
      iconPrefix: 'mdi-'
    };
    var faIcons = function faIcons() {
      var faIconPrefix = config && config.defaultIconComponent ? '' : 'fa-';
      return {
        sizes: {
          "default": null,
          'is-small': null,
          'is-medium': faIconPrefix + 'lg',
          'is-large': faIconPrefix + '2x'
        },
        iconPrefix: faIconPrefix,
        internalIcons: {
          information: 'info-circle',
          alert: 'exclamation-triangle',
          'alert-circle': 'exclamation-circle',
          'chevron-right': 'angle-right',
          'chevron-left': 'angle-left',
          'chevron-down': 'angle-down',
          'eye-off': 'eye-slash',
          'menu-down': 'caret-down',
          'menu-up': 'caret-up',
          'close-circle': 'times-circle'
        }
      };
    };
    var getIcons = function getIcons() {
      var icons = {
        mdi: mdiIcons,
        fa: faIcons(),
        fas: faIcons(),
        far: faIcons(),
        fad: faIcons(),
        fab: faIcons(),
        fal: faIcons(),
        'fa-solid': faIcons(),
        'fa-regular': faIcons(),
        'fa-light': faIcons(),
        'fa-thin': faIcons(),
        'fa-duotone': faIcons(),
        'fa-brands': faIcons()
      };
      if (config && config.customIconPacks) {
        icons = merge(icons, config.customIconPacks, true);
      }
      return icons;
    };
    var getIcons$1 = getIcons;

    var script$1 = {
        name: 'BIcon',
        props: {
            type: [String, Object],
            component: String,
            pack: String,
            icon: String,
            size: String,
            customSize: String,
            customClass: String,
            both: Boolean // This is used internally to show both MDI and FA icon
        },
        computed: {
            iconConfig() {
                const allIcons = getIcons$1();
                return allIcons[this.newPack]
            },
            iconPrefix() {
                if (this.iconConfig && this.iconConfig.iconPrefix) {
                    return this.iconConfig.iconPrefix
                }
                return ''
            },
            /**
            * Internal icon name based on the pack.
            * If pack is 'fa', gets the equivalent FA icon name of the MDI,
            * internal icons are always MDI.
            */
            newIcon() {
                return `${this.iconPrefix}${this.getEquivalentIconOf(this.icon)}`
            },
            newPack() {
                return this.pack || config.defaultIconPack
            },
            newType() {
                if (!this.type) return

                let splitType = [];
                if (typeof this.type === 'string') {
                    splitType = this.type.split('-');
                } else {
                    for (const key in this.type) {
                        if (this.type[key]) {
                            splitType = key.split('-');
                            break
                        }
                    }
                }
                if (splitType.length <= 1) return

                const [, ...type] = splitType;
                return `has-text-${type.join('-')}`
            },
            newCustomSize() {
                return this.customSize || this.customSizeByPack
            },
            customSizeByPack() {
                if (this.iconConfig && this.iconConfig.sizes) {
                    if (this.size && this.iconConfig.sizes[this.size] !== undefined) {
                        return this.iconConfig.sizes[this.size]
                    } else if (this.iconConfig.sizes.default) {
                        return this.iconConfig.sizes.default
                    }
                }
                return null
            },
            useIconComponent() {
                return this.component || config.defaultIconComponent
            }
        },
        methods: {
            /**
            * Equivalent icon name of the MDI.
            */
            getEquivalentIconOf(value) {
                // Only transform the class if the both prop is set to true
                if (!this.both) {
                    return value
                }

                if (this.iconConfig &&
                    this.iconConfig.internalIcons &&
                    this.iconConfig.internalIcons[value]) {
                    return this.iconConfig.internalIcons[value]
                }
                return value
            }
        }
    };

    function render$1(_ctx, _cache, $props, $setup, $data, $options) {
      return (vue.openBlock(), vue.createElementBlock("span", {
        class: vue.normalizeClass(["icon", [$options.newType, $props.size]])
      }, [
        (!$options.useIconComponent)
          ? (vue.openBlock(), vue.createElementBlock("i", {
              key: 0,
              class: vue.normalizeClass([$options.newPack, $options.newIcon, $options.newCustomSize, $props.customClass])
            }, null, 2 /* CLASS */))
          : (vue.openBlock(), vue.createBlock(vue.resolveDynamicComponent($options.useIconComponent), {
              key: 1,
              icon: [$options.newPack, $options.newIcon],
              size: $options.newCustomSize,
              class: vue.normalizeClass([$props.customClass])
            }, null, 8 /* PROPS */, ["icon", "size", "class"]))
      ], 2 /* CLASS */))
    }

    script$1.render = render$1;
    script$1.__file = "src/components/icon/Icon.vue";

    function _objectWithoutPropertiesLoose(source, excluded) {
      if (source == null) return {};
      var target = {};
      var sourceKeys = Object.keys(source);
      var key, i;
      for (i = 0; i < sourceKeys.length; i++) {
        key = sourceKeys[i];
        if (excluded.indexOf(key) >= 0) continue;
        target[key] = source[key];
      }
      return target;
    }

    function _objectWithoutProperties(source, excluded) {
      if (source == null) return {};
      var target = _objectWithoutPropertiesLoose(source, excluded);
      var key, i;
      if (Object.getOwnPropertySymbols) {
        var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
        for (i = 0; i < sourceSymbolKeys.length; i++) {
          key = sourceSymbolKeys[i];
          if (excluded.indexOf(key) >= 0) continue;
          if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
          target[key] = source[key];
        }
      }
      return target;
    }

    var _excluded = ["style", "class", "id"];
    var CompatFallthroughMixin = {
      inheritAttrs: false,
      props: {
        compatFallthrough: {
          type: Boolean,
          "default": function _default() {
            return config.defaultCompatFallthrough;
          }
        }
      },
      computed: {
        rootAttrs: function rootAttrs() {
          return this.compatFallthrough ? {
            "class": this.$attrs["class"],
            style: this.$attrs.style,
            id: this.$attrs.id
          } : {};
        },
        fallthroughAttrs: function fallthroughAttrs() {
          if (this.compatFallthrough) {
            var _this$$attrs = this.$attrs;
              _this$$attrs.style;
              _this$$attrs["class"];
              _this$$attrs.id;
              var rest = _objectWithoutProperties(_this$$attrs, _excluded);
            return rest;
          } else {
            return this.$attrs;
          }
        }
      }
    };

    var script = {
        name: 'BMenuItem',
        components: {
            [script$1.name]: script$1
        },
        mixins: [CompatFallthroughMixin, MenuItemContainerMixin],
        inject: {
            parent: {
                from: 'BMenuItemContainer',
                default: null
            }
        },
        // deprecated, to replace with default 'value' in the next breaking change
        model: {
            prop: 'active',
            event: 'update:active'
        },
        props: {
            label: String,
            active: Boolean,
            expanded: Boolean,
            disabled: Boolean,
            iconPack: String,
            icon: String,
            animation: {
                type: String,
                default: 'slide'
            },
            tag: {
                type: String,
                default: 'a',
                validator: (value) => {
                    return config.defaultLinkTags.indexOf(value) >= 0
                }
            },
            ariaRole: {
                type: String,
                default: ''
            },
            size: {
                type: String,
                default: 'is-small'
            }
        },
        emits: ['update:active', 'update:expanded'],
        data() {
            return {
                newActive: this.active,
                newExpanded: this.expanded
            }
        },
        computed: {
            ariaRoleMenu() {
                return this.ariaRole === 'menuitem' ? this.ariaRole : null
            }
        },
        watch: {
            active(value) {
                this.newActive = value;
            },
            expanded(value) {
                this.newExpanded = value;
            }
        },
        methods: {
            onClick(event) {
                if (this.disabled) return
                const menu = this.getMenu();
                this.reset(this.parent, menu);
                this.newExpanded = this.$props.expanded || !this.newExpanded;
                this.$emit('update:expanded', this.newExpanded);
                if (menu && menu.activable) {
                    this.newActive = true;
                    this.$emit('update:active', this.newActive);
                }
            },
            reset(parent, menu) {
                if (parent == null) {
                    return
                }
                parent.menuItems.forEach((item) => {
                    if (item !== this) {
                        this.reset(item, menu);
                        if (!parent.$data._isMenu || (parent.$data._isMenu && parent.accordion)) {
                            item.newExpanded = false;
                            item.$emit('update:expanded', item.newActive);
                        }
                        if (menu && menu.activable) {
                            item.newActive = false;
                            item.$emit('update:active', item.newActive);
                        }
                    }
                });
            },
            getMenu() {
                let parent = this.$parent;
                while (parent && !parent.$data._isMenu) {
                    parent = parent.$parent;
                }
                return parent
            }
        },
        mounted() {
            if (this.parent) {
                this.parent.appendMenuItem(this);
            }
        },
        beforeUnmount() {
            if (this.parent) {
                this.parent.removeMenuItem(this);
            }
        }
    };

    const _hoisted_1 = ["role"];
    const _hoisted_2 = { key: 1 };

    function render(_ctx, _cache, $props, $setup, $data, $options) {
      const _component_b_icon = vue.resolveComponent("b-icon");

      return (vue.openBlock(), vue.createElementBlock("li", vue.mergeProps({ role: $options.ariaRoleMenu }, _ctx.rootAttrs), [
        (vue.openBlock(), vue.createBlock(vue.resolveDynamicComponent($props.tag), vue.mergeProps(_ctx.fallthroughAttrs, {
          class: {
                    'is-active': $data.newActive,
                    'is-expanded': $data.newExpanded,
                    'is-disabled': $props.disabled,
                    'icon-text': $props.icon,
                },
          onClick: _cache[0] || (_cache[0] = $event => ($options.onClick($event)))
        }), {
          default: vue.withCtx(() => [
            ($props.icon)
              ? (vue.openBlock(), vue.createBlock(_component_b_icon, {
                  key: 0,
                  icon: $props.icon,
                  pack: $props.iconPack,
                  size: $props.size
                }, null, 8 /* PROPS */, ["icon", "pack", "size"]))
              : vue.createCommentVNode("v-if", true),
            ($props.label)
              ? (vue.openBlock(), vue.createElementBlock("span", _hoisted_2, vue.toDisplayString($props.label), 1 /* TEXT */))
              : vue.renderSlot(_ctx.$slots, "label", {
                  key: 2,
                  expanded: $data.newExpanded,
                  active: $data.newActive
                })
          ]),
          _: 3 /* FORWARDED */
        }, 16 /* FULL_PROPS */, ["class"])),
        vue.createCommentVNode(" sub menu items "),
        (_ctx.$slots.default)
          ? (vue.openBlock(), vue.createBlock(vue.Transition, {
              key: 0,
              name: $props.animation,
              persisted: ""
            }, {
              default: vue.withCtx(() => [
                vue.withDirectives(vue.createElementVNode("ul", null, [
                  vue.renderSlot(_ctx.$slots, "default")
                ], 512 /* NEED_PATCH */), [
                  [vue.vShow, $data.newExpanded]
                ])
              ]),
              _: 3 /* FORWARDED */
            }, 8 /* PROPS */, ["name"]))
          : vue.createCommentVNode("v-if", true)
      ], 16 /* FULL_PROPS */, _hoisted_1))
    }

    script.render = render;
    script.__file = "src/components/menu/MenuItem.vue";

    var use = function use(plugin) {
      if (typeof window !== 'undefined' && window.Vue) {
        window.Vue.use(plugin);
      }
    };

    // use `name` to register a Functional Component which will become unresolvable
    // in production build due to name mangling.
    var registerComponent = function registerComponent(Vue, component, name) {
      Vue.component(name || component.name, component);
    };

    var Plugin = {
      install: function install(Vue) {
        registerComponent(Vue, script$3);
        // explicit `name` is needed to avoid name mangling of
        // Functional Component in production
        registerComponent(Vue, script$2, 'BMenuList');
        registerComponent(Vue, script);
      }
    };
    use(Plugin);

    exports.BMenu = script$3;
    exports.BMenuItem = script;
    exports.BMenuList = script$2;
    exports.default = Plugin;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
